third_party/nvidia/lib/Dialect/NVWS/Transforms/InsertAref.cpp

```diff
@@ -0,0 +1,398 @@
+    }
+  };
+  auto [tensorType, desc] = getTensorTypeAndDesc(descOp);
+  auto encoding = getEncodingFromDescriptor(descOp, tensorType, desc);
+  auto shapePerCTA = getShapePerCTA(encoding, tensorType.getShape());
+  return product(shapePerCTA) *
+         tensorType.getElementType().getIntOrFloatBitWidth() / 8;
+}
+
+void createNVWSDescriptorLoadOp(OpBuilder &builder, Operation *ttDescLoadOp,
+                                Value dataBuf, Partition *producerPartition,
+                                WarpSchedule &schedule, Location loc) {
+  auto txCount = getTxCount(ttDescLoadOp);
+  if (auto descLoad = dyn_cast<triton::DescriptorLoadOp>(ttDescLoadOp)) {
+    auto newDescLoad = builder.create<triton::nvws::DescriptorLoadOp>(
+        loc, descLoad.getDesc(), descLoad.getIndices(), txCount, dataBuf,
+        descLoad.getCache(), descLoad.getEvict());
+    newDescLoad->setAttrs(descLoad->getAttrs());
+    schedule.insert(producerPartition, newDescLoad);
+  } else if (auto descGather =
+                 dyn_cast<triton::DescriptorGatherOp>(ttDescLoadOp)) {
+    auto newDescGather = builder.create<triton::nvws::DescriptorGatherOp>(
+        loc, descGather.getDesc(), descGather.getXOffsets(),
+        descGather.getYOffset(), txCount, dataBuf);
+    newDescGather->setAttrs(descGather->getAttrs());
+    schedule.insert(producerPartition, newDescGather);
+  } else {
+    llvm_unreachable("unknown descriptor op.");
+  }
+}
+
+Value mkConstant(OpBuilder &builder, Location loc, int value, int width,
+                 Partition *partition, WarpSchedule &schedule) {
+  auto constValue = builder.create<arith::ConstantIntOp>(loc, value, width);
+  schedule.insert(partition, constValue);
+  return constValue;
+}
+
+StageCluster getStageClusterForProducer(Value producedValue) {
+  if (isDescLoadAndAlloc(producedValue) ||
+      isGlobalLoadAndAlloc(producedValue)) {
+    auto alloc = producedValue.getDefiningOp<LocalAllocOp>();
+    auto loadOp = alloc.getSrc().getDefiningOp();
+    return getStageCluster(loadOp);
+  }
+  return getStageCluster(producedValue.getDefiningOp());
+}
+
+SmallVector<Operation *> createArefPut(PartitionBuilder &builder,
+                                       ArefCreateOp aref, std::string arefTag,
+                                       ProducedValueInfo producedValue,
+                                       Partition *producerPartition,
+                                       WarpSchedule &schedule) {
+  auto loc = producedValue.result.getLoc();
+  auto arefBufType = cast<MemDescType>(aref.getOperand(0).getType());
+  Value result = producedValue.result;
+  auto dataBufType = getDataMemDescType(arefBufType, true);
+  StageCluster stageCluster = getStageClusterForProducer(result);
+
+  SmallVector<Type> buffers{dataBufType};
+
+  auto putEnterOp = builder.createInto<ArefPutEnterOp>(
+      *producerPartition, stageCluster, buffers, aref,
+      mkConstant(builder, loc, 0, 32, producerPartition, schedule));
```

Comment:
can you add a helper for this on PartitionBuilder?

Reply:
Sorry, I meant the `mkConstant` method.

Reply:
Actually, PartitionBuilder already has `intCst` helper. Now using that and removed `mkConstant`.